### PR TITLE
Updated the TreeFlatteningRebuildingVisitor to work iteratively.

### PR DESCRIPTION
**Pull request by jwomeara**

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuilder.java [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+     * If the original node has no children, we will simply use the RebuildingVisitor to copy the node.
+     *
+     * @param node
+     *            the node to copy
+     * @return the copied node
+     */
+    private JexlNode rebuildNode(JexlNode node) {
+        JexlNode newNode;
+        if (node.jjtGetNumChildren() == 0) {
+            newNode = RebuildingVisitor.copy(node);
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuilder.java]: Do we need this portion of the if?  I think the else part will work when numChildren == 0...

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuilder.java [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+        // if the parent is a NOT node, keep the reference
+        if (!removeReferences || parent instanceof ASTNotNode || JexlASTHelper.dereference(node) instanceof ASTAssignment
+                        || QueryPropertyMarker.instanceOf(node, null)) {
+            return node;
+        }
+        // if this is a (reference -> reference expression) combo with a single child, remove the reference and reference expression
+        else if (node.jjtGetNumChildren() == 1 && node.jjtGetChild(0) instanceof ASTReferenceExpression) {
+            ASTReferenceExpression refExp = (ASTReferenceExpression) node.jjtGetChild(0);
+            if (refExp.jjtGetNumChildren() == 1) {
+                return refExp.jjtGetChild(0);
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuilder.java]: Don't we need to fix up the tree here (reset children of parent and reset the parent reference)? I don't see that being done by the caller above.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuildingVisitor.java [hunk 3]
```diff
@@ -1,49 +1,77 @@
...
-import org.apache.commons.jexl2.parser.JexlNodes;
-import org.apache.commons.jexl2.parser.ParserTreeConstants;
-import org.apache.log4j.Logger;
 
 /**
- * This will flatten ands and ors. If requested this will also remove reference expressions and references where possible. NOTE: If you remove reference
- * expressions and references, this will adversely affect the jexl evaluation of the query.
+ * This is a wrapper for the TreeFlatteningRebuilder which allows the TreeFlatteningRebuilder to be run as a Jexl node visitor.
  */
 public class TreeFlatteningRebuildingVisitor extends RebuildingVisitor {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuildingVisitor.java]: Are we going to deprecate this version?

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuildingVisitor.java [hunk 4]
```diff
@@ -52,246 +80,276 @@ public TreeFlatteningRebuildingVisitor(boolean removeReferences) {
...
     }
     
     @Override
-    public Object visit(ASTOrNode node, Object data) {
-        ASTOrNode orNode = JexlNodes.newInstanceOfType(node);
-        orNode.image = node.image;
-        
-        return flattenTree(node, orNode, data);
+    public Object visit(ASTJexlScript node, Object data) {
+        return treeFlatteningRebuilder.flattenTree(node);
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuildingVisitor.java]: Won't this take care of the entire tree?  Why do we need all of the methods below.  In fact to we actually need this one?

File: warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java [hunk 5]
```diff
@@ -1458,6 +1462,60 @@ public static Double getNodeSelectivity(Set<String> fieldNames, ShardQueryConfig
...
+                for (JexlNode child : children(node)) {
+                    if (child != null) {
+                        if (child.jjtGetParent() == null) {
+                            if (failHard)
+                                throw new RuntimeException("Failed to validate lineage: Tree included a child with a null parent.");
+                            else
+                                log.error("Failed to validate lineage: Tree included a child with a null parent.");
+                            
+                            result = false;
+                        } else if (!child.jjtGetParent().equals(node)) {
```
[the remark this answers is not in the document]
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java]: you'r right.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/QueryPropertyMarkerVisitor.java [hunk 6]
```diff
@@ -103,36 +106,50 @@ public Object visit(ASTOrNode node, Object data) {
     
     @Override
     public Object visit(ASTAndNode node, Object data) {
-        if (node.jjtGetNumChildren() > 1) {
-            // if this is an and node with multiple children, and it is
-            // the first one we've found, it is our potential candidate
-            if (data == null) {
-                List<JexlNode> siblingNodes = new ArrayList<>();
+        // if this is an and node with multiple children, and it is
```
[the remark this answers is not in the document]
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/QueryPropertyMarkerVisitor.java]: It's still kind of true.  I'll update the empty comment below to add some additional context.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuilder.java [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.commons.jexl2.parser.JexlNodes.children;
+
+/**
+ * This will flatten ands and ors. If requested this will also remove reference expressions and references where possible. NOTE: If you remove reference
+ * expressions and references, this will adversely affect the jexl evaluation of the query.
```
[the remark this answers is not in the document]
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuilder.java]: sure.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuilder.java [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+                newNode = node;
+            }
+            
+            // if we still have nodes to evaluate
+            if (!postOrderStack.isEmpty()) {
+                
+                // if the original node's parent is NOT the next one on the parent stack,
+                // then this is a new parent node. add it to the parent stack, and add a new list of children.
+                // otherwise, add this node to the existing parent's list of children
+                if (!node.jjtGetParent().equals(parentStack.peek())) {
```
[the remark this answers is not in the document]
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuilder.java]: yup.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuildingVisitor.java (hunk 4 above)
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuildingVisitor.java]: That was my initial thought.  Then I ran a build + tests and found that some classes are calling the visitor like so:

TreeFlatteningRebuildingVisitor visitor = new TreeFlatteningRebuildingVisitor(true);
someNode.jjtAccept(visitor, null);

There were a fair amount of classes calling the visitor this way.

**Pull request merged**